### PR TITLE
Staking/pools endpoint

### PR DESCRIPTION
**Pull request by fragosti**
- Adds the 0x-event-pipeline image to the docker-compose file for development. On the first run, it will fail because the postgres db isn't ready yet, but will keep on attempting to run migrations until the DB is ready.
- Refactor the code to dependency inject the db `Connection` instead of making it a global. 
- Add all the scaffolding etc... for the staking endpoint.

Returns a response like this on Kovan at /staking/pools:

```
{
  "currentEpoch": {
    "epochId": 5,
    "epochStart": {
      "blockNumber": 14941779,
      "txHash": "0xb854dae9dbd0e63094bf1ac116a024da020c5676318f8eb87048a525a5b5c725",
      "timestamp": "2019-11-21T01:17:24.000Z"
    }
  },
  "approximateNextEpoch": {
    "epochId": 6,
    "epochStart": {
      "blockNumber": 14970579,
      "timestamp": "2019-11-26T01:17:24.000Z"
    }
  },
  "stakingPools": [
    {
      "poolId": "1",
      "operatorAddress": "0x5409ed021d9299bf6814279a6a1411a7e866a631",
      "createdAt": {
        "blockNumber": 14491738,
        "txHash": "0xea30da4f4a5ccb209fe7861c71b0c365684af076924a23e48a9028c8b7e13e5b"
      },
      "metaData": {
        "isVerified": false
      },
      "currentEpochStats": {
        "poolId": "1",
        "zrxStaked": 6001,
        "operatorShare": 4e-06,
        "approximateStakeRatio": 0,
        "makerAddresses": [
          "0x5409ed021d9299bf6814279a6a1411a7e866a631"
        ],
        "protocolFeesGeneratedInEth": 0
      },
      "nextEpochStats": {
        "poolId": "1",
        "zrxStaked": 6001,
        "operatorShare": 4e-06,
        "approximateStakeRatio": 0,
        "makerAddresses": [
          "0x5409ed021d9299bf6814279a6a1411a7e866a631"
        ],
        "protocolFeesGeneratedInEth": 0
      }
    },
    {
      "poolId": "2",
      "operatorAddress": "0x6ecbe1db9ef729cbe972c83fb886247691fb6beb",
      "createdAt": {
        "blockNumber": 14576829,
        "txHash": "0x4d587b256b712a456cca2297d599460942444f0cf04774fe48d380b06e89bfb0"
      },
      "metaData": {
        "isVerified": false
      },
      "currentEpochStats": {
        "poolId": "2",
        "zrxStaked": 600,
        "operatorShare": 2e-06,
        "approximateStakeRatio": 0,
        "makerAddresses": [
          "0x6ecbe1db9ef729cbe972c83fb886247691fb6beb"
        ],
        "protocolFeesGeneratedInEth": 0
      },
      "nextEpochStats": {
        "poolId": "2",
        "zrxStaked": 600,
        "operatorShare": 2e-06,
        "approximateStakeRatio": 0,
        "makerAddresses": [
          "0x6ecbe1db9ef729cbe972c83fb886247691fb6beb"
        ],
        "protocolFeesGeneratedInEth": 0
      }
    },
    {
      "poolId": "3",
      "operatorAddress": "0xe36ea790bc9d7ab70c55260c66d52b1eca985f84",
      "createdAt": {
        "blockNumber": 14730024,
        "txHash": "0x7cdeab4f2dbc5e231db5b28593891943c516908af8335f99bc92168e04e6aceb"
      },
      "metaData": {
        "isVerified": false
      },
      "currentEpochStats": {
        "poolId": "3",
        "zrxStaked": 0,
        "operatorShare": 0,
        "approximateStakeRatio": 0,
        "makerAddresses": [],
        "protocolFeesGeneratedInEth": 0
      },
      "nextEpochStats": {
        "poolId": "3",
        "zrxStaked": 0,
        "operatorShare": 0.49,
        "approximateStakeRatio": 0,
        "makerAddresses": [
          "0xe36ea790bc9d7ab70c55260c66d52b1eca985f84"
        ],
        "protocolFeesGeneratedInEth": 0
      }
    },
    {
      "poolId": "4",
      "operatorAddress": "0x3998a82afec0bbde9021fcf16c753b3cbb6a78b2",
      "createdAt": {
        "blockNumber": 14768235,
        "txHash": "0x0bea3f8bf9e35c00313fc8126f2997c3eda95bef33a24e7e71b5c78d9390ace0"
      },
      "metaData": {
        "isVerified": false
      },
      "currentEpochStats": {
        "poolId": "4",
        "zrxStaked": 0,
        "operatorShare": 0,
        "approximateStakeRatio": 0,
        "makerAddresses": [],
        "protocolFeesGeneratedInEth": 0
      },
      "nextEpochStats": {
        "poolId": "4",
        "zrxStaked": 0,
        "operatorShare": 9.5e-05,
        "approximateStakeRatio": 0,
        "makerAddresses": [
          "0x3998a82afec0bbde9021fcf16c753b3cbb6a78b2"
        ],
        "protocolFeesGeneratedInEth": 0
      }
    },
    {
      "poolId": "5",
      "operatorAddress": "0x3998a82afec0bbde9021fcf16c753b3cbb6a78b2",
      "createdAt": {
        "blockNumber": 14768243,
        "txHash": "0x0463f2b864b14262377b4edfd42e48ff0c514a95a138269084aa1e52aadf609b"
      },
      "metaData": {
        "isVerified": false
      },
      "currentEpochStats": {
        "poolId": "5",
        "zrxStaked": 100,
        "operatorShare": 0.7,
        "approximateStakeRatio": 0,
        "makerAddresses": [
          "0x3998a82afec0bbde9021fcf16c753b3cbb6a78b2"
        ],
        "protocolFeesGeneratedInEth": 0
      },
      "nextEpochStats": {
        "poolId": "5",
        "zrxStaked": 100,
        "operatorShare": 0.7,
        "approximateStakeRatio": 0,
        "makerAddresses": [
          "0x3998a82afec0bbde9021fcf16c753b3cbb6a78b2"
        ],
        "protocolFeesGeneratedInEth": 0
      }
    },
    {
      "poolId": "6",
      "operatorAddress": "0x3998a82afec0bbde9021fcf16c753b3cbb6a78b2",
      "createdAt": {
        "blockNumber": 14769171,
        "txHash": "0x8687710e59c44f0632e9a64f67c03e30dc5e52dac198bd04a26628bec3321e90"
      },
      "metaData": {
        "isVerified": false
      },
      "currentEpochStats": {
        "poolId": "6",
        "zrxStaked": 0,
        "operatorShare": 0,
        "approximateStakeRatio": 0,
        "makerAddresses": [],
        "protocolFeesGeneratedInEth": 0
      },
      "nextEpochStats": {
        "poolId": "6",
        "zrxStaked": 0,
        "operatorShare": 0.999999,
        "approximateStakeRatio": 0,
        "makerAddresses": [
          "0x3998a82afec0bbde9021fcf16c753b3cbb6a78b2"
        ],
        "protocolFeesGeneratedInEth": 0
      }
    },
    {
      "poolId": "7",
      "operatorAddress": "0x5409ed021d9299bf6814279a6a1411a7e866a631",
      "createdAt": {
        "blockNumber": 14797004,
        "txHash": "0x99a283cb02f925cc03d20998f0c565e388fc0740c5f7ace360b3cb07f4d7a5eb"
      },
      "metaData": {
        "isVerified": false
      },
      "currentEpochStats": {
        "poolId": "7",
        "zrxStaked": 0,
        "operatorShare": 0,
        "approximateStakeRatio": 0,
        "makerAddresses": [],
        "protocolFeesGeneratedInEth": 0
      },
      "nextEpochStats": {
        "poolId": "7",
        "zrxStaked": 0,
        "operatorShare": 1,
        "approximateStakeRatio": 0,
        "makerAddresses": [
          "0x5409ed021d9299bf6814279a6a1411a7e866a631"
        ],
        "protocolFeesGeneratedInEth": 0
      }
    },
    {
      "poolId": "8",
      "operatorAddress": "0x5409ed021d9299bf6814279a6a1411a7e866a631",
      "createdAt": {
        "blockNumber": 14797017,
        "txHash": "0xb6e261156d6a4a01ced813662e305cfd485d487710f66b259fe7e50009e7db87"
      },
      "metaData": {
        "isVerified": false
      },
      "currentEpochStats": {
        "poolId": "8",
        "zrxStaked": 0,
        "operatorShare": 0,
        "approximateStakeRatio": 0,
        "makerAddresses": [],
        "protocolFeesGeneratedInEth": 0
      },
      "nextEpochStats": {
        "poolId": "8",
        "zrxStaked": 0,
        "operatorShare": 0,
        "approximateStakeRatio": 0,
        "makerAddresses": [
          "0x5409ed021d9299bf6814279a6a1411a7e866a631"
        ],
        "protocolFeesGeneratedInEth": 0
      }
    }
  ]
}

```

### CODE REVIEW

File: src/services/staking_data_service.ts [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+        const nextEpochPoolStatsMap = utils.arrayToMapWithId(nextEpochPoolStats, 'poolId');
+        return pools.map(pool => ({
+            ...pool,
+            currentEpochStats: currentEpochPoolStatsMap[pool.poolId],
+            nextEpochStats: nextEpochPoolStatsMap[pool.poolId],
+        }));
+    }
+}
+
+const currentEpocStatsQuery = `
```
Comment (alexkroeger) [inline src/services/staking_data_service.ts]:
```suggestion
const currentEpochStatsQuery = `
```

File: src/types.ts [hunk 2]
```diff
@@ -52,3 +52,95 @@ export interface AddedRemovedUpdate {
...
+    isVerified: boolean;
+    logoUrl?: string;
+    location?: string;
+    bio?: string;
+    websiteUrl?: string;
+    name?: string;
+}
+
+export interface Pool {
+    poolId: number;
```
Comment (alexkroeger) [inline src/types.ts]: Not a fun of this being specified as a number. The sequence is not meaningful.

File: src/types.ts [hunk 3]
```diff
@@ -52,3 +52,95 @@ export interface AddedRemovedUpdate {
...
+    total_staked?: string;
+    share_of_stake?: string;
+    protocol_fees?: string;
+    total_protocol_fees?: string;
+    share_of_fees?: string;
+    approximate_stake_ratio?: string;
+}
+
+export interface EpochPoolStats {
+    poolId: number;
```
Comment (alexkroeger) [inline src/types.ts]: Same comment as above

File: src/services/staking_data_service.ts [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        const nextEpochPoolStatsMap = utils.arrayToMapWithId(nextEpochPoolStats, 'poolId');
+        return pools.map(pool => ({
+            ...pool,
+            currentEpochStats: currentEpochPoolStatsMap[pool.poolId],
+            nextEpochStats: nextEpochPoolStatsMap[pool.poolId],
+        }));
+    }
+}
+
+// TODO: current epoch query does not calculate the stake ratio
```
Comment (alexkroeger) [inline src/services/staking_data_service.ts]: https://github.com/0xProject/0x-event-pipeline/pull/10

File: src/services/staking_data_service.ts [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    public async getCurrentEpochAsync(): Promise<Epoch> {
+        const rawEpoch: RawEpoch | undefined = _.head(await this._connection.query(`SELECT * FROM staking.current_epoch;`));
+        if (!rawEpoch) {
+            throw new Error('Could not find the current epoch.');
+        }
+        const epoch = stakingUtils.getEpochFromRaw(rawEpoch);
+        return epoch;
+    }
+    public async getNextEpochAsync(): Promise<Epoch> {
+        const rawEpoch: RawEpoch | undefined = _.head(await this._connection.query(`SELECT
```
Comment (alexkroeger) [inline src/services/staking_data_service.ts]: In general I think it's worth separating the queries from these blocks of code and refer to an object here.

File: src/types.ts (hunk 2 above)
Comment (fragosti) [reply src/types.ts]: Not sure what you mean.

**Comment by fragosti** [comment]
@alexkroeger I just used `approximateStakeRatio` for both.

**Pull request merged**